src/client/components/filter-menu/time-filter-menu/fixed-time-tab.tsx

```diff
@@ -40,46 +50,83 @@ export interface FixedTimeTabProps {
 }
 
 export interface FixedTimeTabState {
-  start: Date;
-  end: Date;
+  start?: Date;
+  end?: Date;
   shift: string;
+  activeMonthStartDate?: Date;
+  hoverTimeRange?: TimeRange;
+  selectionSet?: boolean;
+  dateRangePickers: List<any>;
+  timeRanges: List<any>;
+  numTimeRanges: int;
 }
 
 export class FixedTimeTab extends React.Component<FixedTimeTabProps, FixedTimeTabState> {
-
   initialState = (): FixedTimeTabState => {
     const { essence, timekeeper, dimension: { name } } = this.props;
     const shift = essence.timeShift.toJS();
-
     const timeFilter = essence.getEffectiveFilter(timekeeper).clauseForReference(name);
     if (timeFilter && timeFilter instanceof FixedTimeFilterClause && !timeFilter.values.isEmpty()) {
-      const { start, end } = timeFilter.values.get(0);
-      return { start, end, shift };
+      if (timeFilter.values.size === 1) {
+        const { start, end } = timeFilter.values.get(0);
+        return { start, end, shift, activeMonthStartDate: null, hoverTimeRange: null, selectionSet: false, dateRangePickers: List(), timeRanges: List(), numTimeRanges: 1 };
+      } else {
+        const { start, end } = timeFilter.values.get(0);
+        let dateRanges = List();
+        for (let i = 0; i < timeFilter.values.size; i++) {
+          dateRanges = dateRanges.push([timeFilter.values.get(i).start, timeFilter.values.get(i).end]);
+        }
+        return { start, end, shift, activeMonthStartDate: null, hoverTimeRange: null, selectionSet: false,
+          dateRangePickers: List(), timeRanges: dateRanges, numTimeRanges: timeFilter.values.size };
+      }
     }
-    return { start: null, end: null, shift };
+    return { start: null, end: null, shift, activeMonthStartDate: null, hoverTimeRange: null, selectionSet: false, dateRangePickers: List(), timeRanges: List(), numTimeRanges: 1 };
   }
 
-  onStartChange = (start: Date) => this.setState({ start });
+  componentWillMount() {
```

Comment:
I don't like bringing logic to parent component. It already does a lot. Extract to different component.

Reply:
So do you want the calendar portion of the render to get moved into its own component? I don't quite understand what you mean by this.

Reply:
So I see it like that: Now `FixedTimeTab` is responsible for managing time filter that is currently edited. Creating new, validation and displaying controls for editing (`DateRangePicker`). Now, we're introducing another "mode" of editing filter - not single picker, but list of editable pickers. And that list should be separate component. `FixedTimeTab` should display `DateRangePicker` (with calendar) and `MultiplaDataRangePicker` (or some better name :)). It should hold state for determining in which mode it is and should pass callbacks for settings it's internal state (`onDateRangeChange` or something).